File: dashboard/backend/app/platform/surface_registry.py
"""Platform surface registry — M3.1 classification (Public / Client / CEO)."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

SurfaceTarget = Literal["public", "client", "ceo"]
ShellKind = Literal["public", "mission_control", "transitional"]

_REPO_ROOT = Path(__file__).resolve().parents[3]
_REGISTRY_PATH = _REPO_ROOT / "platform" / "surface_registry.json"


@dataclass(frozen=True)
class SurfaceRoute:
    path: str
    label: str
    target: SurfaceTarget
    current_shell: ShellKind
    capabilities: tuple[str, ...]
    overlap: tuple[SurfaceTarget, ...]
    migration_note: str

    def matches(self, pathname: str) -> bool:
        if self.path == pathname:
            return True
        pattern = re.sub(r":(\w+)", r"[^/]+", self.path)
        regex = "^" + pattern.replace("/", r"\/") + "$"
        return bool(re.match(regex, pathname))


def load_registry() -> dict[str, Any]:
    data = json.loads(_REGISTRY_PATH.read_text(encoding="utf-8"))
    if not isinstance(data, dict):
        raise ValueError("surface_registry.json must be an object")
    return data


def parse_routes(data: dict[str, Any] | None = None) -> list[SurfaceRoute]:
    raw = data or load_registry()
    routes: list[SurfaceRoute] = []
    for item in raw.get("routes") or []:
        if not isinstance(item, dict):
            continue
        routes.append(
            SurfaceRoute(
                path=str(item.get("path") or ""),
                label=str(item.get("label") or ""),
                target=item.get("target") or "public",
                current_shell=item.get("current_shell") or "public",
                capabilities=tuple(item.get("capabilities") or []),
                overlap=tuple(item.get("overlap") or []),
                migration_note=str(item.get("migration_note") or ""),
            )
        )
    return routes


def routes_for_target(target: SurfaceTarget) -> list[SurfaceRoute]:
    return [r for r in parse_routes() if r.target == target or target in r.overlap]


def classify_path(pathname: str) -> SurfaceRoute | None:
    best: SurfaceRoute | None = None
    best_len = -1
    for route in parse_routes():
        if route.matches(pathname) and len(route.path) > best_len:
            best = route
            best_len = len(route.path)
    return best


def client_nav_paths() -> list[str]:
    nav = load_registry().get("navigation") or {}
    paths = nav.get("client_nav_paths") or ["/projects"]
    return [str(p) for p in paths]


def resolve_navigation_surface(pathname: str) -> SurfaceTarget:
    for prefix in client_nav_paths():
        if pathname == prefix or pathname.startswith(f"{prefix}/"):
            return "client"
    if pathname == "/":
        return "ceo"
    if pathname == "/products":
        return "public"
    for prefix in ceo_mc_prefixes():
        if pathname == prefix or (prefix != "/" and pathname.startswith(f"{prefix}/")):
            return "ceo"
    return "public"


def ceo_mc_prefixes() -> list[str]:
    """Paths that use Mission Control chrome today (AppShell MC_PREFIXES alignment)."""
    prefixes: set[str] = set()
    for route in parse_routes():
        if route.current_shell == "mission_control":
            base = route.path.split("/:")[0].rstrip("/") or "/"
            prefixes.add(base if base != "" else "/")
    return sorted(prefixes, key=lambda p: (p != "/", p))

File: dashboard/frontend/app/lib/surfaceRegistry.ts
/**
 * M3.1 Surface Registry — Public / Client / CEO classification.
 * Source of truth: dashboard/platform/surface_registry.json
 */

import registry from "../../../platform/surface_registry.json";

export type SurfaceTarget = "public" | "client" | "ceo";
export type ShellKind = "public" | "mission_control" | "transitional";

export type SurfaceRoute = {
  path: string;
  label: string;
  target: SurfaceTarget;
  current_shell: ShellKind;
  capabilities: string[];
  overlap?: SurfaceTarget[];
  migration_note?: string;
};

export type SurfaceRegistry = {
  version: string;
  routes: SurfaceRoute[];
  kernel_layers: string[];
  slice_gate: string[];
};

export const SURFACE_REGISTRY = registry as SurfaceRegistry;

function patternToRegex(path: string): RegExp {
  const escaped = path
    .split("/")
    .map((seg) => (seg.startsWith(":") ? "[^/]+" : seg.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")))
    .join("/");
  return new RegExp(`^${escaped}$`);
}

export function classifyPath(pathname: string): SurfaceRoute | undefined {
  let best: SurfaceRoute | undefined;
  let bestLen = -1;
  for (const route of SURFACE_REGISTRY.routes) {
    if (patternToRegex(route.path).test(pathname) && route.path.length > bestLen) {
      best = route;
      bestLen = route.path.length;
    }
  }
  return best;
}

export function routesForTarget(target: SurfaceTarget): SurfaceRoute[] {
  return SURFACE_REGISTRY.routes.filter(
    (r) => r.target === target || (r.overlap ?? []).includes(target),
  );
}

/** Mission Control chrome prefixes — mirrors AppShell until M3.2 wires registry directly. */
export function missionControlPrefixes(): string[] {
  const prefixes = new Set<string>();
  for (const route of SURFACE_REGISTRY.routes) {
    if (route.current_shell === "mission_control") {
      const base = route.path.split("/:")[0] || "/";
      prefixes.add(base === "" ? "/" : base);
    }
  }
  return Array.from(prefixes);
}

export function isMissionControlRoute(pathname: string): boolean {
  if (pathname === "/") return true;
  if (pathname === "/products") return false;
  return missionControlPrefixes().some(
    (p) => pathname === p || (p !== "/" && pathname.startsWith(`${p}/`)),
  );
}

export const SLICE_GATE = SURFACE_REGISTRY.slice_gate;
